### DISCUSSION TITLE
Make InodeCache to work on FreeBSD (ask for feedback)

### DISCUSSION THREAD

**Discussion by olegsidorkin** [ccache/ccache#1228] [Development]
I'm trying to make InodeCache to work under FreeBSD OS. It has different pthread_mutex implementation that destroys all mutexes when the last file mapping containing these mutexes is unmapped.
<details><summary>details from man libthr on FreeBSD</summary>
<p>

>      The following sysctl MIBs affect the operation of the library: 
>
>      kern.ipc.umtx_vnode_persistent       By default, a shared lock backed by 
>                                           a mapped file in memory is
>                                           automatically destroyed on the last
>                                           unmap of the corresponding file's
>                                           page, which is allowed by POSIX.
>                                           Setting the sysctl to 1 makes such a
>                                           shared lock object persist until the
>                                           vnode is recycled by the Virtual 
>                                           File System.  Note that in case file
>                                           is not opened and not mapped, the
>                                           kernel might recycle it at any
>                                           moment, making this sysctl less
>                                           useful than it sounds.
</p>
</details>

So it's impossible to make currently implemented locking method to run reliably on FreeBSD. My idea is to switch to ```fcntl``` based locking, It's a POSIX standard so it looks more universal. Patch that works for me is [in my branch](https://github.com/ccache/ccache/commit/7037311ad86e90c421e3df637005ff19a3aeb661). 

Now the question is: Is it worth upstreaming and how. ```fcntl```  locks are probably a bit slower than pthread mutexes and FreeBSD share is quite low (and other non-linux bases OSes shares are even lower)

My variants are:

1. Do nothing. These changes can live in my own branch, and, probably be commited downstream.
    Pros: We need to do nothing.
    Cons: Downstream has a bit more things to test

2. Make it a compile-time option (hide behind defines or templates).
    Pros:
    - We have a method to fallback to
    Cons:
    - We have to maintain and test it separately
3. Replace pthread mutexes with ```fcntl``` everywhere.
    Pros:
    - Single method that is widely supported
    - No strange bytes in the file. Only data is there.
    Cons:
    - Maybe a bit slower

**Comment by jrosdahl** [comment]
Are you sure that fcntl locks would be an appropriate alternative to pthread mutexes for accessing memory mapped files? Correct me if I'm wrong, but I think that fcntl locks don't imply memory barriers, so shared memory is not guaranteed to be in sync between CPUs when synchronizing via fcntl locks. But perhaps it works in practice?

So the alternative would be to use fcntl locks + open/read/write/close calls, not mmap.

@liljenzin: Any thoughts here?

**Reply by olegsidorkin** [reply]
> Are you sure that fcntl locks would be an appropriate alternative to pthread mutexes for accessing memory mapped files? Correct me if I'm wrong, but I think that fcntl locks don't imply memory barriers, so shared memory is not guaranteed to be in sync between CPUs when synchronizing via fcntl locks. But perhaps it works in practice?

AFAIK, [tdb in samba used this scheme at least for a while](https://github.com/samba-team/samba/blame/master/lib/tdb/docs/mutex.txt) with no extra barriers (and they switched to mmaped mutexes due to fcntl slowness). If not, a memory barriers should be added inside protected code. 

Uncontended fcntl is reported to be about two times slower than pthread mutex. This is the reason I'm trying to be careful about the default implementation.

**Reply by liljenzin** [reply]
When running on modern machines with dozens of cores it is probably not feasible to lock the entire cache with a single lock and hold it during the process lifetime, since that would effectively serialize the build process in the ideal case where everything is perfectly cached. Add obtaining and releasing the lock every time the cache is accessed is probably also a problem since each time ccache is invoked it might need to do thousands of lookups in just a few milliseconds. The current implementations avoids contention by splitting up the cache in thousands of buckets where each bucket is guarded by its own mutex, thus allowing concurrent access to different parts of the cache. Not sure how you would implement something similar with fcntl without making it very slow. (Taking into account that memory will have to be manually syncronized in case of fcntl as @jrosdahl already pointed out.)

Regarding data consistency, if ccache crashes or is prematurely terminated by an external signal it might be in a middle of a write that leaves cache entries in inconsistent state. At least on Linux that supports robust mutexes, the next process that tries to lock a bucket will then detect that the mutex was left in locked state and will if that was the case discard the potentially corrupt bucket. In case of fcntl the lock will be released when the process holding it dies, but it will not be possible for the next process obtaining the lock to see why the lock was released by the process holding it before. That is also something to consider if changing to fcntl.

**Reply by olegsidorkin** [reply]
Locks design wasn't changed: each fcntl call still locks the bucket with a bytes-range lock. Multiple ccache instances can still access the cache if they do use the same bucket.

Checking the bucket for consistency is a really good point. Thanks. It could be fixed by using some kind of flag that indicates that the write is in progress.

I'm completely ok to keep this implementation in my fork (because it's the only implementation that works for me) - probably it's the best solution for now.

Thanks again for feedback.

**Reply by liljenzin** [reply]
> Checking the bucket for consistency is a really good point. Thanks. It could be fixed by using some kind of flag that indicates that the write is in progress.

When you get that working you have essentially implemented your own mutexes and don't need file locks in the first please, right? But in general you can't count on things like strict memory ordering that will make writes to the flag visible before writes to the guarded data when read by other cores, unless you also implement memory barriers that preserve memory ordering. And no, it is not guaranteed to work without barriers in practice unless you write assembler code for it. On Intel/AMD some instructions are write through while other instructions are write back and without special care you don't know what instructions the compiler will decide to use. And on ARM the ordering is weaker in general.

It is not that I'm against fixing this on platforms that don't have pthread_mutex, but that it all seems to work when you test it on your local machine doesn't imply that it always will work on a larger scale.

**Reply by olegsidorkin** [reply]
> When you get that working you have essentially implemented your own mutexes and don't need file locks in the first please, right?

Yes, it will behave like a mutex, and no: fcntl is another way to talk to the kernel (to suspend process while resource is locked and unlock the resource if process dies).

>But in general you can't count on things like strict memory ordering that will make writes to the flag visible before writes to the guarded 
> data when read by other cores, unless you also implement memory barriers that preserve memory ordering

I think fcntl lock already provides enough barriers (since tdb uses it for few years, and I don't think it's possible to check that the region is locked inside the kernel without barriers). But I can have per bucket atomic variable to read and write (I need a variable anyway to prevent dirty reads in case of crashes). 

> And no, it is not guaranteed to work without barriers in practice unless you write assembler code for it. On Intel/AMD some instructions 
>are write through while other instructions are write back and without special care you don't know what instructions the compiler will 
>decide to use. And on ARM the ordering is weaker in general.

I hope, standard C++11 atomics already have required barriers in all implementations, and I don't have to insert assembly instructions.

> It is not that I'm against fixing this on platforms that don't have pthread_mutex, but that it all seems to work when you test it on your
> local machine doesn't imply that it always will work on a larger scale.

Yes, it should be reviewed carefully before merging to any project repository, even under the knob.

**Reply by liljenzin** [reply]
> I hope, standard C++11 atomics already have required barriers in all implementations, and I don't have to insert assembly instructions.

Sure you can use std::atomic to implement your own mutex:

```
std::atomic<pid_t>& lock = { some address }; 
 
// obtain lock
pid_t mypid = getpid(); 
while(true) { 
  pid_t expected = 0; 
  if (lock.compare_exchange_weak(expected, mypid)) 
    break; 
  // lock is held by a process with pid {expected} 
  sched_yield(); 
} 
 
// lock is now obtained 
 
if (!lock.compare_exchange_strong(mypid, 0)) { 
  // Something is broken... 
  assert(false); 
} 
 
// lock is now released 
```
To handle the case with stale locks from crashed processes you could check in the while loop if the holding pid is never changing, and perhaps after a few seconds check if the holder is still alive. If not obtain the lock yourself, wipe the cached bucket and life is good.

**Reply by olegsidorkin** [reply]
I thought about spinlocks before using fcntl. I didn't like them because I had to guess the other process state.
It can be really tricky because containers can have overlapping pids and there's quite impossible to check the other process state in different container. Timeouts are the other sort of heuristics: they are magic constants in code and they are always too big to make things fast and too small to debug code with GDB.

**Reply by liljenzin** [reply]
> I thought about spinlocks before using fcntl. I didn't like them because I had to guess the other process state. It can be really tricky because containers can have overlapping pids and there's quite impossible to check the other process state in different container.

If you use containers with overlapping pids they shouldn't share the same cache file and probably not other files as well. Saving the own pid in lock files to let other processes remove the locks if the pid is no longer alive, is a commonly used trick potentially used in many parts of a larger build system. If pids are not unique within used file trees you are basically violating common assumptions. 

> Timeouts are the other sort of heuristics: they are magic constants in code and they are always too big to make things fast and too small to debug code with GDB.

Not an issue as long as you keep looping while the other process is still alive.

You should implement it as a double loop where you loop over sched_yield() in the inner loop and check if pid is alive in the outer loop. The reason for a double loop is that sched_yield() isn't taking cpu if other processes want to run while checking pids does take cpu.

Then if we do the math we will find the probability that we have to loop many rounds over sched_yield() is very small, because the number of buckets is large, because the time the holder of a mutex will keep it is small and because sched_yield() will always move the waiting process last in the run queue.

**Reply by olegsidorkin** [reply]
Ok, with spinlocks: https://github.com/olegsidorkin/ccache/tree/Inodes-cache-use-spinlocks
(no ccache processes are seen in top during digikam compilation with filled cache)

**Reply by jrosdahl** [reply]
Thanks. I've now started to look at the PR and think about the solution in more detail.

@liljenzin wrote:

> If you use containers with overlapping pids they shouldn't share the same cache file and probably not other files as well. Saving the own pid in lock files to let other processes remove the locks if the pid is no longer alive, is a commonly used trick potentially used in many parts of a larger build system. If pids are not unique within used file trees you are basically violating common assumptions.

I agree with this in principle. However, in practice people could (and some most likely will) start containers with a mounted shared cache directory. If they then don't have a `/run/user/<pid>` directory in the containers, the inode cache in `$CCACHE_DIR/tmp` will be shared as well, and we'll risk PID collisions. I don't think that it's reasonable to just document that the cache directory should not be shared between containers when using the inode cache, so I don't feel comfortable enabling the inode cache by default if we risk PID collision.

I would be OK with keeping the pthread implementation for Linux (leaving inode caching enabled by default) and providing the spin lock alternative for other UNIXes in combination with making inode caching disabled by default. A unified solution that could be enabled by default would of course be better, but I'm not sure what it would be.

**Reply by liljenzin** [reply]
> I agree with this in principle. However, in practice people could (and some most likely will) start containers with a mounted shared cache directory. If they then don't have a `/run/user/<pid>` directory in the containers, the inode cache in `$CCACHE_DIR/tmp` will be shared as well, and we'll risk PID collisions. I don't think that it's reasonable to just document that the cache directory should not be shared between containers when using the inode cache, so I don't feel comfortable enabling the inode cache by default if we risk PID collision.
> 
> I would be OK with keeping the pthread implementation for Linux (leaving inode caching enabled by default) and providing the spin lock alternative for other UNIXes in combination with making inode caching disabled by default. A unified solution that could be enabled by default would of course be better, but I'm not sure what it would be.

That stale locks are being detected should be rare, even if possible with false positives in containers, simply because there is no good reason for a healthy process to hold a lock for very long. Perhaps it would be good enough to replace the entire cache file with a fresh one instead of a wiping a single bucket if a lock is not released or changes owner within reasonable time (a few seconds)? Then it would turn into a performance issue if it happens too often (which it shouldn't), but not to a potentially corrupt cache. This as a solution for systems that don't have robust mutexes, while keeping pthread mutexes on Linux where they seem to work well?

**Reply by olegsidorkin** [reply]
> I agree with this in principle. However, in practice people could (and some most likely will) start containers with a mounted shared cache directory. If they then don't have a `/run/user/<pid>` directory in the containers, the inode cache in `$CCACHE_DIR/tmp` will be shared as well, and we'll risk PID collisions. I don't think that it's reasonable to just document that the cache directory should not be shared between containers when using the inode cache, so I don't feel comfortable enabling the inode cache by default if we risk PID collision.

Similar thoughts led me to the initial fcntl locks solution.

> I would be OK with keeping the pthread implementation for Linux (leaving inode caching enabled by default) and providing the spin lock alternative for other UNIXes in combination with making inode caching disabled by default. A unified solution that could be enabled by default would of course be better, but I'm not sure what it would be.

Fcntl locks are universal enough (however could be slower).

**Reply by liljenzin** [reply]
Note that there are other platforms that support pthread_mutex but don't support robust mutexes. On such platforms ccache will just print an error message telling the user to remove the cache file manually to fix the problem. Unless we have got reports about this being a nuisance I guess this isn't happening very often in real life. Perhaps almost never. And as a general rule of thumb we should optimize the code for the normal case (where mutexes or spin locks are considerably faster than fcntl locks) and not for rare error cases as long as the latter can be handled in a safe way that can't cause inconsistency. So wouldn't it be good enough to just print an error message if this happens with spin locks as well?

**Reply by jrosdahl** [reply]
@liljenzin wrote:
>  On such platforms ccache will just print an error message telling the user to remove the cache file manually to fix the problem. Unless we have got reports about this being a nuisance I guess this isn't happening very often in real life. 

Note that the message will only be printed to the log file, which a typical user won't see, so I don't think we can draw conclusions from the absence of related bug reports.

> Perhaps it would be good enough to replace the entire cache file with a fresh one instead of a wiping a single bucket if a lock is not released or changes owner within reasonable time (a few seconds)? Then it would turn into a performance issue if it happens too often (which it shouldn't), but not to a potentially corrupt cache.

This sounds OK to me.

> This as a solution for systems that don't have robust mutexes, while keeping pthread mutexes on Linux where they seem to work well?

It would be nice not to have two different solutions. Would you be OK with the spin-lock-with-file-replace solution for all systems?

**Reply by liljenzin** [reply]
> It would be nice not to have two different solutions. Would you be OK with the spin-lock-with-file-replace solution for all systems?

If we go that road I think the procedure in the review needs a little tweak though. We should not only check the current lock owner is alive in the outer loop, but also in the inner loop check that the owner isn't changing between iterations, otherwise reset the timer since changed ownership indicates the owners are alive. This should reduce the risk that we reach the timeout just because of contention on a machine with huge number of cores.

**Reply by jrosdahl** [reply]
Yes, restarting the attempt to acquire the lock when an owner change is detected sounds good. However, we can't check if the owner is alive since the process could be in another PID namespace, right?

**Reply by liljenzin** [reply]
> Yes, restarting the attempt to acquire the lock when an owner change is detected sounds good. However, we can't check if the owner is alive since the process could be in another PID namespace, right?

I think we should still check if the process is alive since that should work in the normal case where cache files are not shared between containers using local pids, which would eliminate the risk for unnecessarily wiping the cache on systems where it is safe to not do so.

**Reply by jrosdahl** [reply]
Agreed.

**Reply by jrosdahl** [reply]
Second thoughts: If we reset the timer (or iteration counter) when we detect an alive process with the owner PID, then we risk spinning for a very long time if the process is a long-lived process (say, a daemon or a login shell) in another PID namespace.

**Reply by liljenzin** [reply]
Then skip checking pid alive but still reset the timer if the owner changes (as you already suggested) should be safe?

**Reply by jrosdahl** [reply]
Yep!

**Reply by olegsidorkin** [reply]
Updated pr: now only timeout is checked.
Probably Duration constructor could be made constexpr, but this is for another PR.

**Comment by liljenzin** [comment]
Looks good to me if all tests pass, it compiles on platforms where it compiled before and so on.

**Reply by olegsidorkin** [reply]
Should I send it as a pull request?

**Reply by liljenzin** [reply]
> Should I send it as a pull request?

In my opinion yes, but then it is @jrosdahl that has the final say in this project.

**Reply by jrosdahl** [reply]
Sounds good to me as well.

**Reply by olegsidorkin** [reply]
Ok. I'm going look at it one more time, and then I'll make a PR.

**Reply by olegsidorkin** [reply]
Done (with style fixes): https://github.com/ccache/ccache/pull/1229

**Comment by olegsidorkin** [comment]
Has anybody tested inodes cache on macos (inodes cache unit test is marked as skipped in github actions)?
I don't have a mac around to test it. There should be the same problem with destroyed mutexes if macos has the same libthr implementation as freebsd.
